File: ios/Packages/MauryaEffects/Sources/MauryaEffects/EffectProgramRepository.swift
import Foundation

public protocol EffectProgramStorage: Sendable {
    func read() throws -> Data?
    func replaceAtomically(with data: Data) throws
    @discardableResult func quarantineCorruptFile() throws -> URL?
}

public struct FileEffectProgramStorage: EffectProgramStorage, Sendable {
    public let fileURL: URL

    public init(directoryURL: URL, fileName: String = "effect_programs.json") {
        fileURL = directoryURL.appendingPathComponent(fileName, isDirectory: false)
    }

    public func read() throws -> Data? {
        do { return try Data(contentsOf: fileURL) } catch let error as CocoaError where error.code == .fileReadNoSuchFile { return nil }
    }

    public func replaceAtomically(with data: Data) throws {
        try FileManager.default.createDirectory(at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try data.write(to: fileURL, options: .atomic)
    }

    @discardableResult public func quarantineCorruptFile() throws -> URL? {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return nil }
        let destination = fileURL.deletingLastPathComponent()
            .appendingPathComponent("\(fileURL.lastPathComponent).corrupt-\(UUID().uuidString)")
        try FileManager.default.moveItem(at: fileURL, to: destination)
        return destination
    }
}

public struct EffectProgramRecord: Equatable, Sendable {
    public let program: EffectProgram
    public let revision: String

    public init(program: EffectProgram, revision: String) {
        self.program = program
        self.revision = revision
    }
}

public enum EffectProgramRecovery: Equatable, Sendable {
    case none
    case createdDefaults
    case replacedCorruptFile(quarantineURL: URL?)
}

public actor EffectProgramRepository {
    public typealias Clock = @Sendable () -> Int64
    public typealias IDGenerator = @Sendable () -> String

    private let storage: any EffectProgramStorage
    private let clock: Clock
    private let makeID: IDGenerator
    private var values: [EffectProgram]
    public nonisolated let recovery: EffectProgramRecovery

    public init(
        storage: any EffectProgramStorage,
        defaults: [EffectProgram]? = nil,
        clock: @escaping Clock = { Int64(Date().timeIntervalSince1970 * 1_000) },
        makeID: @escaping IDGenerator = { UUID().uuidString }
    ) throws {
        self.storage = storage
        self.clock = clock
        self.makeID = makeID
        let now = clock()
        let fallback = try defaults ?? Self.androidExamples(now: now)
        let initialValues: [EffectProgram]
        let initialRecovery: EffectProgramRecovery
        if let data = try storage.read() {
            do {
                initialValues = try EffectProgramTransfer.decodeRepository(data, now: now, makeID: makeID)
                initialRecovery = .none
            } catch {
                let quarantined = try? storage.quarantineCorruptFile()
                initialValues = fallback
                initialRecovery = .replacedCorruptFile(quarantineURL: quarantined ?? nil)
            }
            try storage.replaceAtomically(with: EffectProgramTransfer.encodeRepository(initialValues))
        } else {
            initialValues = fallback
            initialRecovery = .createdDefaults
            try storage.replaceAtomically(with: EffectProgramTransfer.encodeRepository(fallback))
        }
        values = initialValues
        recovery = initialRecovery
    }

    public func list() throws -> [EffectProgramRecord] { try values.map(Self.record) }
    public func programs() -> [EffectProgram] { values }

    @discardableResult public func upsert(_ program: EffectProgram, expectedRevision: String? = nil) throws -> EffectProgramRecord {
        var next = values
        let index = next.firstIndex { $0.id == program.id }
        if let index {
            if let expectedRevision, try EffectProgramTransfer.revision(of: next[index]) != expectedRevision {
                throw EffectProgramError.revisionConflict
            }
            next[index] = try EffectProgramCompiler.normalise(program, now: clock())
        } else {
            guard expectedRevision == nil else { throw EffectProgramError.revisionConflict }
            guard next.count < EffectProgramTransfer.maximumPrograms else { throw EffectProgramError.tooManyPrograms }
            next.append(try EffectProgramCompiler.normalise(program, now: clock()))
        }
        try commit(next)
        return try Self.record(next[index ?? (next.count - 1)])
    }

    @discardableResult public func copy(id: String) throws -> EffectProgramRecord {
        guard let source = values.first(where: { $0.id == id }) else { throw EffectProgramError.programNotFound(id) }
        guard values.count < EffectProgramTransfer.maximumPrograms else { throw EffectProgramError.tooManyPrograms }
        let now = clock()
        let duplicate = EffectProgram(
            id: makeID(), nameZh: "\(source.nameZh) 副本", nameJa: "\(source.nameJa) コピー",
            workspaceJSON: source.workspaceJSON, astJSON: source.astJSON, astSHA256: source.astSHA256,
            blockCount: source.blockCount, estimatedDurationMilliseconds: source.estimatedDurationMilliseconds,
            createdAt: now, updatedAt: now, editorSchema: source.editorSchema, programSchema: source.programSchema,
            sourceKind: source.sourceKind, scriptSource: source.scriptSource
        )
        var next = values; next.append(duplicate)
        try commit(next)
        return try Self.record(duplicate)
    }

    @discardableResult public func delete(id: String, expectedRevision: String? = nil) throws -> [EffectProgramRecord] {
        guard let index = values.firstIndex(where: { $0.id == id }) else { return try list() }
        if let expectedRevision, try EffectProgramTransfer.revision(of: values[index]) != expectedRevision {
            throw EffectProgramError.revisionConflict
        }
        var next = values; next.remove(at: index)
        try commit(next)
        return try next.map(Self.record)
    }

    public func exportProgram(id: String) throws -> Data {
        guard let program = values.first(where: { $0.id == id }) else { throw EffectProgramError.programNotFound(id) }
        return try EffectProgramTransfer.exportSingle(program)
    }

    public func exportAll() throws -> Data { try EffectProgramTransfer.exportBundle(values) }

    public func previewImport(_ data: Data) throws -> EffectImportPreview {
        try EffectProgramTransfer.preview(data, existingIDs: Set(values.map(\.id)), now: clock, makeID: makeID)
    }

    @discardableResult public func applyImport(_ preview: EffectImportPreview, strategy: EffectImportConflictStrategy) throws
        -> [EffectProgramRecord]
    {
        var next = values
        for importedValue in preview.programs {
            let imported = try EffectProgramCompiler.normalise(importedValue, now: importedValue.updatedAt)
            if let index = next.firstIndex(where: { $0.id == imported.id }) {
                switch strategy {
                case .skip: continue
                case .overwrite: next[index] = imported
                case .copy:
                    let now = clock()
                    next.append(
                        EffectProgram(
                            id: makeID(), nameZh: "\(imported.nameZh) 副本", nameJa: "\(imported.nameJa) コピー",
                            workspaceJSON: imported.workspaceJSON, astJSON: imported.astJSON, astSHA256: imported.astSHA256,
                            blockCount: imported.blockCount, estimatedDurationMilliseconds: imported.estimatedDurationMilliseconds,
                            createdAt: now, updatedAt: now, editorSchema: imported.editorSchema, programSchema: imported.programSchema,
                            sourceKind: imported.sourceKind, scriptSource: imported.scriptSource
                        ))
                }
            } else {
                next.append(imported)
            }
        }
        guard next.count <= EffectProgramTransfer.maximumPrograms else { throw EffectProgramError.tooManyPrograms }
        guard Set(next.map(\.id)).count == next.count else { throw EffectProgramError.duplicateProgramID("import") }
        try commit(next)
        return try next.map(Self.record)
    }

    private func commit(_ next: [EffectProgram]) throws {
        let data = try EffectProgramTransfer.encodeRepository(next)
        try storage.replaceAtomically(with: data)
        values = next
    }

    private static func record(_ program: EffectProgram) throws -> EffectProgramRecord {
        try EffectProgramRecord(program: program, revision: EffectProgramTransfer.revision(of: program))
    }

    private static func androidExamples(now: Int64) throws -> [EffectProgram] {
        let workspaces = [
            (
                "example-rgb", "红→绿→蓝", "赤→緑→青",
                ##"{"blocks":{"languageVersion":0,"blocks":[{"type":"maurya_start","id":"example-rgb-start","next":{"block":{"type":"maurya_set_color","id":"example-rgb-red","fields":{"TARGET":"ALL","COLOR":"#ff0000"},"next":{"block":{"type":"maurya_wait","id":"example-rgb-wait","fields":{"DURATION":500,"UNIT":"MS"},"next":{"block":{"type":"maurya_fade","id":"example-rgb-green","fields":{"TARGET":"ALL","COLOR":"#00ff00","DURATION":1500},"next":{"block":{"type":"maurya_fade","id":"example-rgb-blue","fields":{"TARGET":"ALL","COLOR":"#0000ff","DURATION":1500}}}}}}}}}}]}}"##
            ),
            (
                "example-rainbow", "无限彩虹", "無限レインボー",
                ##"{"blocks":{"languageVersion":0,"blocks":[{"type":"maurya_start","id":"example-rainbow-start","next":{"block":{"type":"maurya_forever","id":"example-rainbow-loop","inputs":{"DO":{"block":{"type":"maurya_adjust_hsv","id":"example-rainbow-adjust","fields":{"TARGET":"ALL","H":2,"S":0,"V":0},"next":{"block":{"type":"maurya_wait","id":"example-rainbow-wait","fields":{"DURATION":50,"UNIT":"MS"}}}}}}}}}]}}"##
            ),
        ]
        return try workspaces.map { id, zh, ja, workspace in
            try EffectProgramCompiler.normalise(
                EffectProgram(
                    id: id, nameZh: zh, nameJa: ja, workspaceJSON: workspace, createdAt: now, updatedAt: now, sourceKind: .blocks),
                now: now
            )
        }
    }
}

File: android/app/src/main/java/com/example/peacock/feature/effects/EffectProgramRepository.kt
package com.example.peacock.feature.effects

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

class EffectProgramRepository(context: Context) {
    private val file = File(context.filesDir, "effect_programs.json")
    private val preferences = context.getSharedPreferences("effect_program_repository", Context.MODE_PRIVATE)

    @Synchronized
    fun load(): List<EffectProgram> {
        if (!file.exists()) return examples().also {
            saveAll(it)
            markBundledExamplesInstalled()
        }
        return runCatching {
            val array = JSONArray(file.readText())
            val decoded = (0 until array.length()).map { decode(array.getJSONObject(it)) }
            installBundledExamples(migratePrograms(decoded))
        }.getOrElse { examples().also(::saveAll) }
    }

    @Synchronized
    fun upsert(program: EffectProgram): List<EffectProgram> {
        val next = load().toMutableList()
        val index = next.indexOfFirst { it.id == program.id }
        if (index >= 0) next[index] = program else {
            require(next.size < 50) { "最多保存50个灯效程序" }
            next += program
        }
        saveAll(next)
        return next
    }

    @Synchronized
    fun delete(id: String): List<EffectProgram> =
        load().filterNot { it.id == id }.also(::saveAll)

    @Synchronized
    fun exportProgram(id: String): String {
        val program = load().firstOrNull { it.id == id }
            ?: error("找不到灯效程序 / エフェクトプログラムが見つかりません")
        return EffectProgramTransfer.exportSingle(program)
    }

    @Synchronized
    fun exportAll(): String = EffectProgramTransfer.exportBundle(load())

    @Synchronized
    fun previewImport(text: String): EffectImportPreview =
        EffectProgramTransfer.preview(text, load().map(EffectProgram::id).toSet())

    @Synchronized
    fun applyImport(
        preview: EffectImportPreview,
        strategy: EffectImportConflictStrategy,
    ): List<EffectProgram> {
        val current = load().toMutableList()
        preview.programs.forEach { imported ->
            val existing = current.indexOfFirst { it.id == imported.id }
            when {
                existing < 0 -> current += imported
                strategy == EffectImportConflictStrategy.SKIP -> Unit
                strategy == EffectImportConflictStrategy.OVERWRITE -> current[existing] = imported
                else -> current += imported.copy(
                    id = UUID.randomUUID().toString(),
                    nameZh = "${imported.nameZh} 副本",
                    nameJa = "${imported.nameJa} コピー",
                    createdAt = System.currentTimeMillis(),
                    updatedAt = System.currentTimeMillis(),
                )
            }
        }
        require(current.size <= 50) {
            "导入后会超过50个程序 / インポート後に50件を超えます"
        }
        saveAll(current)
        return current
    }

    private fun saveAll(programs: List<EffectProgram>) {
        val temp = File(file.parentFile, "${file.name}.tmp")
        temp.writeText(JSONArray().apply { programs.forEach { put(encode(it)) } }.toString())
        if (!temp.renameTo(file)) {
            file.writeText(temp.readText())
            temp.delete()
        }
    }

    private fun examples(): List<EffectProgram> {
        val now = System.currentTimeMillis()
        return listOf(
            example(
                "example-rgb", "红→绿→蓝", "赤→緑→青",
                chain(
                    block("maurya_start"),
                    block("maurya_set_color", mapOf("TARGET" to "ALL", "COLOR" to "#ff0000")),
                    block("maurya_wait", mapOf("DURATION" to 500, "UNIT" to "MS")),
                    block("maurya_fade", mapOf("TARGET" to "ALL", "COLOR" to "#00ff00", "DURATION" to 1500)),
                    block("maurya_fade", mapOf("TARGET" to "ALL", "COLOR" to "#0000ff", "DURATION" to 1500)),
                ), now,
            ),
            example(
                "example-rainbow", "无限彩虹", "無限レインボー",
                chain(
                    block("maurya_start"),
                    loop(block("maurya_adjust_hsv", mapOf("TARGET" to "ALL", "H" to 2, "S" to 0, "V" to 0)),
                        block("maurya_wait", mapOf("DURATION" to 50, "UNIT" to "MS"))),
                ), now,
            ),
            scriptExample(
                "example-nebula-prism",
                "星云棱镜·42灯独立",
                "ネビュラプリズム・42灯独立",
                BuiltinEffectSources.NEBULA_PRISM,
                now,
            ),
        )
    }

    private fun example(id:String, zh:String, ja:String, workspace:String, now:Long):EffectProgram {
        val compiled=EffectCompiler.compile(workspace)
        return EffectProgram(id,zh,ja,workspace,EffectCompiler.canonicalJson(compiled),compiled.astSha256,
            compiled.blockCount,compiled.estimatedDurationMs,now,now,
            editorSchema=EffectProgramSchemas.EDITOR,
            programSchema=EffectProgramSchemas.PROGRAM)
    }

    private fun scriptExample(id: String, zh: String, ja: String, source: String, now: Long): EffectProgram =
        EffectProgramCompiler.normalise(
            EffectProgram(
                id = id,
                nameZh = zh,
                nameJa = ja,
                workspaceJson = "",
                astJson = "",
                astSha256 = "",
                blockCount = 0,
                estimatedDurationMs = null,
                createdAt = now,
                updatedAt = now,
                editorSchema = EffectProgramSchemas.EDITOR,
                programSchema = EffectProgramSchemas.PROGRAM,
                sourceKind = EffectSourceKind.SCRIPT,
                scriptSource = source,
            ),
            now,
        )

    private fun installBundledExamples(programs: List<EffectProgram>): List<EffectProgram> {
        if (preferences.getInt(BUNDLED_EXAMPLES_VERSION_KEY, 0) >= BUNDLED_EXAMPLES_VERSION) {
            return programs
        }
        val existingIds = programs.mapTo(mutableSetOf(), EffectProgram::id)
        val additions = examples().filter { it.id in BUNDLED_ADDITION_IDS && it.id !in existingIds }
        val next = if (programs.size + additions.size <= 50) programs + additions else programs
        if (next !== programs) saveAll(next)
        markBundledExamplesInstalled()
        return next
    }

    private fun markBundledExamplesInstalled() {
        preferences.edit().putInt(BUNDLED_EXAMPLES_VERSION_KEY, BUNDLED_EXAMPLES_VERSION).apply()
    }

    private fun migratePrograms(programs: List<EffectProgram>): List<EffectProgram> {
        val replacements = examples().associateBy { it.id }
        var changed = false
        val migrated = programs.map { program ->
            val replacement = replacements[program.id]
            if (replacement != null && program.programSchema < 3) {
                changed = true
                replacement.copy(createdAt = program.createdAt)
            } else if (
                program.programSchema < EffectProgramSchemas.PROGRAM ||
                program.editorSchema < EffectProgramSchemas.EDITOR
            ) {
                runCatching {
                    changed = true
                    EffectProgramCompiler.normalise(program)
                }.getOrDefault(program)
            } else {
                program
            }
        }
        if (changed) saveAll(migrated)
        return migrated
    }

    private fun chain(vararg blocks:JSONObject):String {
        for(i in 0 until blocks.lastIndex) blocks[i].put("next",JSONObject().put("block",blocks[i+1]))
        return JSONObject().put("blocks",JSONObject().put("languageVersion",0).put("blocks",JSONArray().put(blocks[0]))).toString()
    }
    private fun block(type:String, fields:Map<String,Any> = emptyMap())=JSONObject().put("type",type).put("id",UUID.randomUUID().toString())
        .also { if(fields.isNotEmpty()) it.put("fields",JSONObject(fields)) }
    private fun loop(vararg body:JSONObject):JSONObject {
        for(i in 0 until body.lastIndex) body[i].put("next",JSONObject().put("block",body[i+1]))
        return block("maurya_forever").put("inputs",JSONObject().put("DO",JSONObject().put("block",body[0])))
    }

    private fun encode(p:EffectProgram)=EffectProgramTransfer.encode(p)
    private fun decode(o:JSONObject)=EffectProgramTransfer.decode(o)

    private companion object {
        const val BUNDLED_EXAMPLES_VERSION_KEY = "bundled_examples_version"
        const val BUNDLED_EXAMPLES_VERSION = 1
        val BUNDLED_ADDITION_IDS = setOf("example-nebula-prism")
    }
}
